File: sdk/kotlin-enhanced/src/jvmMain/kotlin/com/flagent/enhanced/cache/EvaluationCache.kt
package com.flagent.enhanced.cache

import com.flagent.client.models.EvalResult
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.concurrent.ConcurrentHashMap

data class CacheKey(
    val flagKey: String?,
    val flagID: Long?,
    val entityID: String?,
    val entityType: String?
) {
    fun toKeyString(): String = "${flagID ?: flagKey}_${entityID}_${entityType}"
}

private data class CachedEntry(
    val result: EvalResult,
    val timestamp: Long
) {
    fun isExpired(ttlMs: Long, currentTime: Long): Boolean =
        (currentTime - timestamp) > ttlMs
}

interface EvaluationCache {
    suspend fun get(key: CacheKey): EvalResult?
    suspend fun put(key: CacheKey, result: EvalResult)
    suspend fun clear()
    suspend fun evictExpired()
}

class InMemoryEvaluationCache(
    private val ttlMs: Long
) : EvaluationCache {
    private val cache = ConcurrentHashMap<String, CachedEntry>()
    private val mutex = Mutex()

    override suspend fun get(key: CacheKey): EvalResult? = mutex.withLock {
        val entry = cache[key.toKeyString()]
        if (entry != null && !entry.isExpired(ttlMs, System.currentTimeMillis())) {
            entry.result
        } else {
            cache.remove(key.toKeyString())
            null
        }
    }

    override suspend fun put(key: CacheKey, result: EvalResult) = mutex.withLock {
        cache[key.toKeyString()] = CachedEntry(result, System.currentTimeMillis())
    }

    override suspend fun clear() = mutex.withLock { cache.clear() }

    override suspend fun evictExpired() = mutex.withLock {
        val currentTime = System.currentTimeMillis()
        cache.entries.filter { it.value.isExpired(ttlMs, currentTime) }.map { it.key }.forEach { cache.remove(it) }
    }
}

File: sdk/flutter-enhanced/lib/src/cache/evaluation_cache.dart
import 'package:flagent_client/flagent_client.dart';

/// Cache key for evaluation results.
///
/// Used to uniquely identify cached evaluation results based on flag and entity.
class CacheKey {
  final String? flagKey;
  final int? flagID;
  final String? entityID;
  final String? entityType;

  const CacheKey({
    this.flagKey,
    this.flagID,
    this.entityID,
    this.entityType,
  });

  String toKeyString() =>
      '${flagID ?? flagKey ?? ''}_${entityID ?? ''}_${entityType ?? ''}';
}

/// Cache interface for evaluation results.
abstract class EvaluationCache {
  Future<EvalResult?> get(CacheKey key);
  Future<void> put(CacheKey key, EvalResult result);
  Future<void> clear();
  Future<void> evictExpired();
}

/// In-memory cache implementation with TTL-based expiration.
class InMemoryEvaluationCache implements EvaluationCache {
  final Map<String, _CachedEntry> _cache = {};
  final int ttlMs;

  InMemoryEvaluationCache(this.ttlMs);

  @override
  Future<EvalResult?> get(CacheKey key) async {
    final entry = _cache[key.toKeyString()];
    if (entry == null) return null;
    if (DateTime.now().millisecondsSinceEpoch - entry.timestamp > ttlMs) {
      _cache.remove(key.toKeyString());
      return null;
    }
    return entry.result;
  }

  @override
  Future<void> put(CacheKey key, EvalResult result) async {
    _cache[key.toKeyString()] = _CachedEntry(
      result: result,
      timestamp: DateTime.now().millisecondsSinceEpoch,
    );
  }

  @override
  Future<void> clear() async {
    _cache.clear();
  }

  @override
  Future<void> evictExpired() async {
    final now = DateTime.now().millisecondsSinceEpoch;
    _cache.removeWhere((_, entry) => now - entry.timestamp > ttlMs);
  }
}

class _CachedEntry {
  final EvalResult result;
  final int timestamp;

  _CachedEntry({required this.result, required this.timestamp});
}
